quic/s2n-quic-core/src/crypto/tls/offload.rs

```diff
@@ -0,0 +1,567 @@
+                {
+                    sessions.push((
+                        new_session,
+                        RemoteContext {
+                            tx,
+                            waker: waker.clone(),
+                            receive_initial: AsyncRequest::empty(),
+                            receive_handshake: AsyncRequest::empty(),
+                            receive_application: AsyncRequest::empty(),
+
+                            can_send_initial: AsyncRequest::empty(),
+                            can_send_handshake: AsyncRequest::empty(),
+                            can_send_application: AsyncRequest::empty(),
+                        },
+                    ))
+                }
+
+                let mut next_sessions = vec![];
+
+                // Make progress on all stored sessions, prioritizing existing sessions over incoming ones
+                for (mut session, mut ctx) in sessions {
+                    match session.poll(&mut ctx) {
+                        Poll::Ready(res) => {
+                            let _ = ctx.tx.unbounded_send(Request::Done(session, res));
+                        }
+                        Poll::Pending => {
+                            next_sessions.push((session, ctx));
+                        }
+                    }
+                }
+                sessions = next_sessions;
+
+                thread::park();
+            }
+        });
+
+        Self {
+            inner,
+            remote_thread_waker: task::Waker::from(Arc::new(ThreadWaker(handle.thread().clone()))),
+            _thread: handle,
+            new_session: tx,
+        }
+    }
+}
+
+struct ThreadWaker(thread::Thread);
+
+impl Wake for ThreadWaker {
+    fn wake(self: Arc<Self>) {
+        self.0.unpark();
+    }
+}
+
+impl<E: tls::Endpoint> tls::Endpoint for OffloadEndpoint<E> {
+    type Session = OffloadSession<<E as tls::Endpoint>::Session>;
+
+    fn new_server_session<Params: s2n_codec::EncoderValue>(
+        &mut self,
+        transport_parameters: &Params,
+    ) -> Self::Session {
+        OffloadSession::new(
+            self.inner.new_server_session(transport_parameters),
+            &mut self.new_session,
+            self.remote_thread_waker.clone(),
+        )
+    }
+
+    fn new_client_session<Params: s2n_codec::EncoderValue>(
+        &mut self,
+        transport_parameters: &Params,
+        server_name: application::ServerName,
+    ) -> Self::Session {
+        OffloadSession::new(
+            self.inner
+                .new_client_session(transport_parameters, server_name),
+            &mut self.new_session,
+            self.remote_thread_waker.clone(),
+        )
+    }
+
+    fn max_tag_length(&self) -> usize {
+        self.inner.max_tag_length()
+    }
+}
+
+#[derive(Debug)]
+pub struct OffloadSession<S: tls::Session> {
+    // Inner is none while remote thread has the session
+    inner: Option<S>,
+    is_poll_done: Option<Result<(), crate::transport::Error>>,
```

Comment:
naming on this seems a little off, since its not a boolean. Is there something better?